subsys/net/lib/coap/coap_client.c

```diff
@@ -0,0 +1,716 @@
+		if (ret < 0) {
+			LOG_ERR("Failed to append block 2 option");
+			goto out;
+		}
+	}
+
+	if (req->payload) {
+		uint16_t payload_len;
+		uint16_t offset;
+
+		/* Blockwise send ongoing, add block1 */
+		if (client->send_blk_ctx.total_size > 0 ||
+		   (client->request.max_len - client->request.offset <
+		   req->len + PAYLOAD_MARKER_SIZE)) {
+
+			if (client->send_blk_ctx.total_size == 0) {
+				enum coap_block_size i = get_largest_block1_size(client);
+
+				coap_block_transfer_init(&client->send_blk_ctx, i, req->len);
+			}
+			ret = coap_append_block1_option(&client->request, &client->send_blk_ctx);
+
+			if (ret < 0) {
+				LOG_ERR("Failed to append block1 option");
+				goto out;
+			}
+		}
+
+		ret = coap_packet_append_payload_marker(&client->request);
+
+		if (ret < 0) {
+			LOG_ERR("Failed to append payload marker to CoAP message");
+			goto out;
+		}
+
+		if (client->send_blk_ctx.total_size > 0) {
+			uint16_t block_in_bytes =
+				coap_block_size_to_bytes(client->send_blk_ctx.block_size);
+
+			payload_len = client->send_blk_ctx.total_size -
+				      client->send_blk_ctx.current;
+			if (payload_len > block_in_bytes) {
+				payload_len = block_in_bytes;
+			}
+			offset = client->send_blk_ctx.current;
+		} else {
+			payload_len = req->len;
+			offset = 0;
+		}
+
+		ret = coap_packet_append_payload(&client->request, req->payload + offset,
+						 payload_len);
+
+		if (ret < 0) {
+			LOG_ERR("Failed to append payload to CoAP message");
+			goto out;
+		}
+
+		if (client->send_blk_ctx.total_size > 0) {
+			coap_next_block(&client->request, &client->send_blk_ctx);
+		}
+	}
+	client->request_tkl = coap_header_get_token(&client->request, client->request_token);
+out:
+	return ret;
+}
+
+
+int coap_client_req(struct coap_client *client, int sock, const struct sockaddr *addr,
+		    struct coap_client_request *req, int retries)
+{
+	int ret;
+
+	if (client->coap_client_recv_active) {
+		return -EAGAIN;
+	}
+
+	if (sock < 0 || req == NULL || req->path == NULL) {
+		return -EINVAL;
+	}
+
+	if (addr != NULL) {
+		memcpy(&client->address, addr, sizeof(*addr));
+		client->socklen = sizeof(client->address);
+	} else {
+		memset(&client->address, 0, sizeof(client->address));
+		client->socklen = 0;
+	}
+
+	if (retries == -1) {
+		client->retry_count = DEFAULT_RETRY_AMOUNT;
+	} else {
+		client->retry_count = retries;
+	}
+
+	ret = coap_client_init_request(client, req);
+	if (ret < 0) {
+		LOG_ERR("Failed to initialize coap request");
+		return ret;
+	}
+
+	ret = coap_client_schedule_poll(client, sock, req);
+	if (ret < 0) {
+		LOG_ERR("Failed to schedule polling");
+		goto out;
+	}
+
+	ret = coap_pending_init(&client->pending, &client->request, &client->address,
+				client->retry_count);
+
+	if (ret < 0) {
+		LOG_ERR("Failed to initialize pending struct");
+		goto out;
+	}
+
+	coap_pending_cycle(&client->pending);
+
+	ret = send_request(sock, client->request.data, client->request.offset, 0, &client->address,
+			   sizeof(client->address));
```

Comment:
This should be `client->socklen` not `sizeof(client->address)` here and everywhere below.

Otherwise, one cannot call `coap_client_req` with the `addr` parameter set to NULL as you intend based on line 299 above.

Reply:
Good catch, fixed now.